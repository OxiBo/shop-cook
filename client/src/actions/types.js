export const SEARCH_RECIPES = "search_recipes";
export const FETCH_RECIPES = "fetch_recipes";
export const IS_LOADING_RECIPES = "is_loading_recipes";
export const IS_LOADING_RECIPE = "is_loading_recipe";
export const FETCH_RECIPE = "fetch_recipe";
export const RECIPES_SEARCH_ERROR = "recipes_search_error";
export const RECIPE_ERROR = "recipe_error";
export const LIKE_RECIPE = "like_recipe";
export const CREATE_SHOPPING_LIST = "create_shopping_list";
// export const UPDATE_SHOPPING_LIST = 'update_shopping_list'
export const FETCH_SHOPPING_LIST = "fetch_shopping_list";
export const IS_LOADING_SHOPPING_LIST = "is_loading_shopping_list";
export const CHANGE_SERVINGS = "change_servings";
export const SIGN_UP = "sign_up";
export const AUTH_ERROR = "auth_error";
export const FETCH_USER ="fetch_user";
export const FETCH_FAV_RECIPES = 'fetch_fav_recipes'
