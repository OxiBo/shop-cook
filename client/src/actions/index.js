import recipes from "../utils/spoonacularAPI";
import {
  SEARCH_RECIPES,
  FETCH_RECIPES,
  IS_LOADING_RECIPES,
  IS_LOADING_RECIPE,
  FETCH_RECIPE,
  RECIPES_SEARCH_ERROR,
  RECIPE_ERROR,
  CREATE_SHOPPING_LIST,
  FETCH_SHOPPING_LIST,
  IS_LOADING_SHOPPING_LIST,
} from "./types";

export const searchRecipes = (name, resultsNumber = 10) => async (dispatch) => {
  try {
    // console.log("dispatch");
    const res = await recipes.get(
      `/complexSearch?query=${name}&number=${resultsNumber}`
    );
    console.log(res);
    dispatch({ type: SEARCH_RECIPES, payload: res.data });
  } catch (error) {
    console.error(error);

    dispatch({ type: RECIPES_SEARCH_ERROR, payload: "Failed to load recipes" });
  }
};

// ???? why
export const fetchRecipes = () => {
  return {
    type: FETCH_RECIPES,
  };
};

export const isLoadingRecipes = () => {
  return {
    type: IS_LOADING_RECIPES,
  };
};

export const isLoadingRecipe = () => {
  return {
    type: IS_LOADING_RECIPE,
  };
};

export const fetchRecipe = (id) => async (dispatch) => {
  try {
    // https://api.spoonacular.com/recipes/511728/information?apiKey=${spoonacularAPI_KEY}
    const res = await recipes.get(`/${id}/information?`);
    // console.log(res);
    const {
      extendedIngredients,
      image,
      readyInMinutes,
      servings,
      sourceName,
      sourceUrl,
      title,
    } = res.data;

    const ingredients = extendedIngredients.reduce(
      (acc, { amount, name, measures, original }) => {
        const parsedIngredient = {
          amount,
          name: name === "seasoning" ? original : name,
          unit: measures.us.unitShort,
          original,
        };
        acc.push(parsedIngredient);
        return acc;
      },
      []
    );
    const recipe = {
      ingredients,
      image,
      readyInMinutes,
      servings,
      sourceName,
      sourceUrl,
      title,
    };
    // console.log(recipe);
    dispatch({ type: FETCH_RECIPE, payload: recipe });
  } catch (error) {
    console.error(error);
    dispatch({ type: RECIPE_ERROR, payload: "Failed to load recipe details" });
  }
};

export const createShoppingList = (ingredients) => async (
  dispatch,
  getState
) => {
  const newShoppingItems = [...ingredients]
  const currentShoppingList = getState().shoppingLists.shoppingList;
  // console.log(currentShoppingList);
  // console.log(ingredients);
  if (currentShoppingList.length) {
    // console.log(ingredients)
    const updatedList = [...currentShoppingList, ...newShoppingItems].reduce((acc, item) => {
      const isExists = acc.findIndex((old) => old.name === item.name && old.unit === item.unit);
      // console.log(isExists)
      if (isExists >= 0) {
        acc[isExists].amount = acc[isExists].amount + item.amount;
        // console.log(acc[isExists])
        // console.log(item)
      } else {
        acc.push(item);
      }
      return acc;
    }, []);
    // console.log(updatedList)
    dispatch({
      type: CREATE_SHOPPING_LIST,
      payload: updatedList,
    });
  } else {
    dispatch({
      type: CREATE_SHOPPING_LIST,
      payload: newShoppingItems
    });
  }
};

export const fetchShoppingList = () => {
  return {
    type: FETCH_SHOPPING_LIST,
  };
};

export const isLoadingShoppingList = () => {
  return {
    type: IS_LOADING_SHOPPING_LIST,
  };
};
