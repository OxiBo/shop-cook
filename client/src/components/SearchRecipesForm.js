import React, { useState } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import SearchStyles from "./styles/SearchStyles";
import Button from "./styles/Button";
import { searchRecipes, isLoadingRecipes } from "../actions";
const SearchBtn = styled.button``;

const SearchRecipesForm = ({ searchRecipes, isLoadingRecipes }) => {
  const [searchValue, setInput] = useState("");
  return (
    <div>
      <SearchStyles
        data-test="recipes-search-form"
        onSubmit={(e) => {
          e.preventDefault();
          console.log(searchValue);
          isLoadingRecipes();
          searchRecipes(searchValue);
          setInput("");
        }}
      >
        <input
          type="text"
          className="search__field"
          placeholder="Search recipe..."
          value={searchValue}
          onChange={(e) => setInput(e.target.value)}
        />
        <Button type="submit">
          {" "}
          <i className="fas fa-search"></i> <span> Search</span>
        </Button>
      </SearchStyles>
    </div>
  );
};

// const mapStateToProps = ({ recipes }) => {
//   return {
//     recipes,
//   };
// };

export default connect(null, { searchRecipes, isLoadingRecipes })(
  SearchRecipesForm
);
