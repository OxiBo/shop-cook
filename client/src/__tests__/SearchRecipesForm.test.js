// https://medium.com/asos-techblog/how-to-test-your-react-redux-application-48d90481a253

import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import SearchRecipesForm from "../components/SearchRecipesForm";
import { fetchRecipes } from "../actions";
const mockStore = configureStore([]);

describe("<SearchRecipesForm />", () => {
  let store, wrapper, componentInstance, searchRecipesForm, shallowWrapper;
  beforeEach(async () => {
    store = mockStore({
      recipes: [],
    });
    store.dispatch = jest.fn();
    wrapper = mount(
      <Provider store={store}>
        <SearchRecipesForm />
      </Provider>
    );

    await wait();
    wrapper.update();
    shallowWrapper = shallow(<SearchRecipesForm store={store} />);
    console.log(shallowWrapper.debug());
    componentInstance = shallowWrapper.dive().instance();
    console.log(componentInstance);
  });
  it("Renders the component properly and matches the snapshot", async () => {
    const searchForm = wrapper.find('[data-test="recipes-search-form"]');
    // console.log(searchForm.debug());
    expect(toJSON(searchForm)).toMatchSnapshot();
  });

  it("submits search query and receives a response from API", async () => {
    wrapper.find("input").simulate("change", { target: { value: "pasta" } });

    // console.log(componentInstance.state());
    // const componentProps = wrapper.instance().props();
    // console.log(componentInstance);
    // mock the router

    wrapper.find("form").simulate("submit",{preventDefault() {}});
    console.log(store.dispatch)
    expect(store.dispatch).toHaveBeenCalledTimes(2); // first call is to fetch recipes, second call is to store fetched recipes in redux store
    expect(store.dispatch.mock.calls).toHaveBeenCalledWith(fetchRecipes("pasta"));
    // await wait(50);
    // wrapper.update();

    //   expect(Router.router.push).toHaveBeenCalled();
    //   expect(Router.router.push).toHaveBeenCalledWith( {"pathname": "/item", "query": {"id": "abc123"}});
  });
});
