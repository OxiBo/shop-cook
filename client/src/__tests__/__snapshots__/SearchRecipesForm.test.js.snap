// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchRecipesForm /> Renders the component properly and matches the snapshot 1`] = `
Array [
  <styled.form
    data-test="recipes-search-form"
    onSubmit={[Function]}
  >
    <form
      className="sc-AxjAm fQhSYg"
      data-test="recipes-search-form"
      onSubmit={[Function]}
    >
      <input
        className="search__field"
        onChange={[Function]}
        placeholder="Search recipe..."
        type="text"
        value=""
      />
      <styled.button
        type="submit"
      >
        <button
          className="sc-AxirZ jA-dooq"
          type="submit"
        >
          Search
        </button>
      </styled.button>
    </form>
  </styled.form>,
  <form
    className="sc-AxjAm fQhSYg"
    data-test="recipes-search-form"
    onSubmit={[Function]}
  >
    <input
      className="search__field"
      onChange={[Function]}
      placeholder="Search recipe..."
      type="text"
      value=""
    />
    <styled.button
      type="submit"
    >
      <button
        className="sc-AxirZ jA-dooq"
        type="submit"
      >
        Search
      </button>
    </styled.button>
  </form>,
]
`;
